galasa-managers-parent/galasa-managers-comms-parent/dev.galasa.mq.manager/src/main/java/dev/galasa/mq/internal/MessageQueueImpl.java

```diff
@@ -0,0 +1,182 @@
+	private boolean started = false;
+	private String currentMethod = new String();
+	private int numberOfMessagesLoggedInThisMethod = 1;
+	
+	private static final Log  logger = LogFactory.getLog(MessageQueueImpl.class);
+	
+	public MessageQueueImpl(String name, MessageQueueManagerImpl qmgr, boolean archive, MQManagerImpl manager) {
+		this.queueName = name;
+		this.qmgr = qmgr;
+		this.archive = archive;
+		this.manager = manager;
+	}
+	
+	/**
+	 * Starts the connection to the queue by obtaining the context from the qmgr
+	 * If we have already been started then just exit
+	 */
+	public void startup() {
+		if(started) {
+			logger.info("Connection to queue: " + this.queueName + " already started");
+			return;
+		}
+		JMSContext context = qmgr.getContext();
+		destination = context.createQueue(QUEUE_PROTOCOL + this.queueName);
+		producer = context.createProducer();
+		consumer = context.createConsumer(destination);
+		this.started = true;
+		logger.info("Connection to queue: " + this.queueName + " complete");
+	}
+
+	@Override
+	public void sendMessage(Message... messages) {
+		for(Message message : messages) {
+			archiveMessage(message,MessageDirection.OUTBOUND);
+			producer.send(destination, message);
+		}
+	}
+
+	@Override
+	public Message getMessage() {
+		Message m = consumer.receive();
+		archiveMessage(m, MessageDirection.INBOUND);
+		return m;
+	}
+
+	@Override
+	public Message getMessage(long timeout) {
+		Message m = consumer.receive(timeout);
+		archiveMessage(m, MessageDirection.INBOUND);
+		return m;
+	}
+
+	@Override
+	public Message getMessageNoWait() {
+		Message m = consumer.receiveNoWait();
+		archiveMessage(m, MessageDirection.INBOUND);
+		return m;
+	}
+	
+	@Override
+	public void clearQueue() {
+		while(consumer.receiveNoWait() != null) {}
+	}
+	
+	/**
+	 * If archival has been set in the annotation for this queue then 
+	 * we archive this message in the RAS in the structure:
+	 * 
+	 * messages
+	 * |
+	 * ---<method name>
+	 *      |
+	 *      ---<queue name>
+	 *           |
+	 *           ---<inbound/outbound>
+	 *               |
+	 *               ---message: <id>
+	 * @param m the message we are archiving
+	 * @param direction the direction in which this message is traveling from the perspective of the test
+	 */
+	private void archiveMessage(Message m, MessageDirection direction) {
+		if(m == null || !archive)
+			return;		
+		byte[] content;
+		try {
+			content = getContentOfMessage(m);
+		}catch (JMSException e) {
+			logger.warn("Unable to retrieve the content of a message while archiving");
+			return;
+		}
+		
+		Path folder = manager.getStoredArtifactRoot()
+							 .resolve(RAS_NAMESPACE)
+							 .resolve(RAS_MESSAGING)
+							 .resolve(getCurrentMethod())
+							 .resolve(this.queueName)
+							 .resolve(direction.toString().toLowerCase())
+							 .resolve("message:" + Integer.toString(numberOfMessagesLoggedInThisMethod));
+		try {
+			Files.write(folder, content, StandardOpenOption.CREATE);
```

Comment:
If the message is text,  then should be written as is.
If the message is binary,   then should be converted to HexDecimal characters so we can read it.  If you intention is to write as is,  then the file SA file should be marked as binary.

Reply:
@Michael-Baylis yes converting to hex characters would be best.  I've updated that